internal/test/scorecard/openshift.go

```diff
@@ -0,0 +1,226 @@
+	clusterVersions := &configv1.ClusterVersionList{}
+	err = openshiftClient.Get().Resource("clusterversions").Do(ctx).Into(clusterVersions)
+	if err != nil {
+		return err
+	}
+	if len(clusterVersions.Items) == 0 {
+		return errors.New("no ClusterVersions found")
+	}
+	clusterVersion := clusterVersions.Items[0]
+	r.Log += fmt.Sprintf("OpenShift version is: %s\n", clusterVersion.Status.Desired.Version)
+	trimmedVer, err := semver.FinalizeVersion(clusterVersion.Status.Desired.Version)
+	if err != nil {
+		return err
+	}
+	version, err := semver.Parse(trimmedVer)
+	if err != nil {
+		return err
+	}
+	// The stable-v1 channel is available on OpenShift 4.12+
+	useStable := false
+	if version.GTE(semver.MustParse("4.12.0")) {
+		useStable = true
+	}
+
+	// Patch the OperatorHub config to enable the default catalog sources
+	hubPatch := `[{"op": "add", "path": "/spec/disableAllDefaultSources", "value": false}]`
+	hub := &configv1.OperatorHub{}
+	err = openshiftClient.Patch(types.JSONPatchType).Resource("operatorhubs").Name("cluster").Body([]byte(hubPatch)).Do(ctx).Into(hub)
+	if err != nil {
+		return err
+	}
+	r.Log += "OperatorHub patched to enable default catalog sources\n"
+
+	// Client for Subscription and ClusterServiceVersion APIs
+	olmClient, err := newRESTClientForGV(config, scheme, &operatorsv1alpha1.SchemeGroupVersion)
+	if err != nil {
+		return err
+	}
+
+	// With the stable-v1 channel, we need to install the operator into one
+	// namespace. This requires us to create the namespace and an OperatorGroup for it.
+	subNamespace := "openshift-operators"
+	channel := "tech-preview"
+	if useStable {
+		subNamespace = "cert-manager-operator"
+		channel = "stable-v1"
+
+		// Client for Namespaces
+		client, err := corev1client.NewForConfig(config)
+		if err != nil {
+			return err
+		}
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "cert-manager-operator",
+			},
+		}
+		_, err = client.Namespaces().Create(ctx, ns, metav1.CreateOptions{})
+		if err != nil && !kerrors.IsAlreadyExists(err) {
+			return err
+		}
+		r.Log += fmt.Sprintf("Created %s namespace\n", subNamespace)
+
+		og := &operatorsv1.OperatorGroup{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "cert-manager-og",
+				Namespace: subNamespace,
+			},
+			Spec: operatorsv1.OperatorGroupSpec{
+				TargetNamespaces: []string{
+					subNamespace,
+				},
+			},
+		}
+
+		// Client for OperatorGroup API
+		ogClient, err := newRESTClientForGV(config, scheme, &operatorsv1.SchemeGroupVersion)
+		if err != nil {
+			return err
+		}
+		err = ogClient.Post().Resource("operatorgroups").Namespace(og.Namespace).Name(og.Name).Body(og).Do(ctx).Into(og)
+		if err != nil && !kerrors.IsAlreadyExists(err) {
+			return err
+		}
+		r.Log += fmt.Sprintf("Created OperatorGroup for %s\n", subNamespace)
+	}
+
+	// Create the Subscription for the cert-manager operator. The namespace and channel
+	// are dependent on the OpenShift version
+	sub := &operatorsv1alpha1.Subscription{
```

Comment:
Just a question here: I have, so far, known `cert-manager` needs to be in `AllNamespaces` mode. What happens if we install it in a single namespace? Wouldn't it be inaccessible outside that namespace anymore?

Reply:
cert-manager is by itself an operator, which watches all namespaces. [The cert-manager Operator for Red Hat OpenShift](https://github.com/openshift/cert-manager-operator) is a sort of meta-operator, that watches a CertManager CRD and installs cert-manager. The operator creates a CertManager CRD on its own with default values, which causes it to install a default cert-manager.

The older version of the operator supported AllNamespaces install mode only. This meant it watched for CertManager CRs in all namespaces. I suppose the authors felt this was unnecessary and limited later versions of the operator to watching a single namespace. In both cases, the resulting cert-manager watches the whole cluster for Issuers, Certificates, etc.

Reply:
Thanks a lot for explaining! I didnt realize the actual cert-manager operand can still have watch the whole cluster for Issues and Certificates. Makes sense now!